File: LeetCode/424_Longest_Repeating_Character_Replacement!!!.py
# Time Complexity: O(26*n) => O(n)

def characterReplacement(s, k):
    count = {}
    result = 0

    l = 0
    for r in range(len(s)):
        count[s[r]] = 1 + count.get(s[r], 0)
        # print(count)

        while (r - l + 1) - max(count.values()) > k:
            count[s[l]] -= 1
            l += 1

        print(max(count.values()))
        print(count)
        result = max(result, r - l + 1)

    return result

def characterReplacement(s,k):
    count = {}
    length = 0
    left = 0
    
    for right in range(len(s)):
        count[s[right]] = 1 + count.get(s[right], 0)

        while (right - left + 1) - max(count.values()) > k:
            count[s[left]] -= 1
            left += 1
        
        length = max(length, (right - left + 1))
    return length

s = "ABAB"
k = 2
print(characterReplacement(s, k))

File: LeetCode/424_Longest_Repeating_Character_Replacement!!!.js
// O(26*n) => O(n)
var characterReplacement = function(s, k) {
    var char = {}
    var left = 0, max_freq = 0, result = 0;

    for (let right = 0; right < s.length; right++) {
        char[s[right]] = 1 + (char[s[right]] || 0)
        max_freq = Math.max(max_freq, char[s[right]])

        while( (right - left + 1) - max_freq > k ) {
            char[s[left]]--
            left++
        }
        result = Math.max(result, right - left + 1)
    }
    return result
}